rfd/0098-registered-openssh-nodes.md

```diff
@@ -0,0 +1,115 @@
+
+## Details
+
+### Registering nodes
+
+A new sub-kind to the `node` resource will be added for registered OpenSSH nodes: `agentless`. The absence of a `node` resource sub-kind will imply that a node is a Teleport agent node, making this change backwards compatible.
+
+#### Registering with `tctl`
+
+`tctl` should not require as many fields to be set when creating nodes. This is an example `node` resource that will work with `tctl create --force` today:
+
+```yaml
+kind: node
+metadata:
+  name: 5da56852-2adb-4540-a37c-80790203f6a9
+spec:
+  addr: 1.2.3.4:22
+  hostname: agentless-node
+version: v2
+```
+
+`tctl` will auto-generate `metadata.name` if it is not already set so users don't have to generate GUIDs themselves if `sub_kind` is `agentless`. Also, if `sub_kind` is set to `agentless`, `spec.public_addr` will not be allowed for registered OpenSSH nodes as it is not needed.
+
+### Agentless CA
+
+For security related reasons that will be discussed below, a new CA will be added called Agentless CA. The Agentless CA will be responsible for generating and signing user certificates that will be used to authenticate with registered OpenSSH nodes. The public key of the Agentless CA will be copied to all registered OpenSSH nodes and configured as `TrustedUserCAKeys` in `sshd_config`.
+
+### RBAC
+
+When OpenSSH nodes are registered currently, RBAC checks for those nodes are not preformed. RBAC logic will have to be updated so RBAC checks for registered OpenSSH nodes are preformed. There are multiple ways to implement this, each with their own advantages and disadvantages.
+
+#### Option 1
+
+Each registered OpenSSH node will have a unique Agentless CA. The CA public key will include information about the node that it was created for, including the unique node name and the node's labels. When `(lib/srv.AuthHandlers).UserKeyAuth` is called to authenticate a node's user certificate, it will check if node information is present in the CA public key. If node information is present, an RBAC check will be preformed using the label information in the CA public key.
+
+Pros:
+
+- No need to lookup node resources to preform RBAC checks
+
+Cons:
+
+- Requires connecting to node to preform RBAC check
+- Updating node labels requires generating and distributing a new host key
+- CA rotation is more complex as every registered OpenSSH node requires a unique CA
+
+#### Option 2
+
+When a user sends a request to a Proxy to connect to a node, the Proxy will attempt to find a node resource by either its hostname or address, whichever the user specified. If the resource exists and has the `agentless` `sub_kind`, an RBAC check will be preformed. If the resource does not exist or isn't an `agentless` node, the connection flow will continue as normal.
+
+Pros:
+
+- No need to connect to node to preform RBAC check
```

Comment:
I don't see how this is relevant. Yes, we don't need to connect to the node to perform RBAC, but eventually, we need to connect to that node anyways. I just don't sound like pros or cons.

Reply:
My original thought was preforming security checks *before* establishing a connection to a node would be safer than preforming security checks *after* establishing a connection to a node. Thinking about it more makes me realize there isn't a huge difference given a realistic threat model, but checking if a connection should be allowed before establishing it should reduce attack surface I'd think.

I think it's a general security principle to delay a sensitive operation as long as possible in code to minimize the impact of a logic bug or vulnerability, though that may not apply here.